packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -27,23 +29,162 @@ import PropTypes from 'prop-types'
  *
  * import {useCategories} from './hooks'
  *
- * const {categories, setCategories} = useCategories()
+ * const {root, findAndModifyFirst, itemsKey, fetchCategoryNode} = useCategories()
  *
  */
 export const CategoriesContext = React.createContext()
-export const CategoriesProvider = ({categories: initialCategories = {}, children}) => {
-    const [categories, setCategories] = useState(initialCategories)
+export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
+    const itemsKey = 'categories'
+    const DEFAULT_ROOT_CATEGORY = 'root'
+    const LOCAL_STORAGE_PREFIX = `pwa-kit-cat-`
+
+    const api = useCommerceAPI()
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: item?.loaded ?? false
+        }))
+    })
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        const findKeys = Object.keys(objToFindBy)
+
+        const rootIsMatch = findKeys.every((key) => isEqual(tree[key], objToFindBy[key]))
+        if (rootIsMatch) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            Object.assign(tree, replacementObj)
+            return tree
```

Comment:
```suggestion
            return {...replacementObj}
```
As discussed [yonder](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/875/files#r1067576151), changing this condition to return a clone is trivial. `findInChildren` is more complex...